receiver/faroreceiver/receiver.go

```
@@ -0,0 +1,212 @@
+	resp.Header().Set("Access-Control-Allow-Headers", "Content-Type")
+
+	// Preflight request
+	if req.Method == http.MethodOptions {
+		resp.WriteHeader(http.StatusOK)
+		return
+	}
+
+	if req.Method != http.MethodPost {
+		r.errorHandler(resp, req, "only POST method is supported", http.StatusMethodNotAllowed)
+		return
+	}
+
+	if req.Header.Get("Content-Type") != "application/json" {
+		r.errorHandler(resp, req, "invalid Content-Type", http.StatusUnsupportedMediaType)
+		return
+	}
+
+	body, err := io.ReadAll(req.Body)
+	if err != nil {
+		r.errorHandler(resp, req, fmt.Sprintf("failed to read request body: %v", err), http.StatusBadRequest)
+		return
+	}
+	defer req.Body.Close()
+
+	var payload faro.Payload
+	if err := json.Unmarshal(body, &payload); err != nil {
+		r.errorHandler(resp, req, fmt.Sprintf("failed to parse Faro payload: %v", err), http.StatusBadRequest)
+		return
+	}
+
+	var errors []string
+
+	if r.nextTraces != nil {
+		traces, err := farotranslator.TranslateToTraces(req.Context(), payload)
+		if err == nil {
+			err = r.nextTraces.ConsumeTraces(req.Context(), traces)
+			if err != nil {
+				errors = append(errors, fmt.Sprintf("failed to push traces: %v", err))
+			}
+		} else {
+			r.settings.Logger.Debug("Faro traces are nil, skipping")
+		}
+	} else {
+		r.settings.Logger.Debug("Traces consumer not registered, skipping")
+	}
+
+	if r.nextLogs != nil {
+		logs, err := farotranslator.TranslateToLogs(req.Context(), payload)
+		if err == nil {
+			err = r.nextLogs.ConsumeLogs(req.Context(), logs)
+			if err != nil {
+				errors = append(errors, fmt.Sprintf("failed to push logs: %v", err))
+			}
+		} else {
+			r.settings.Logger.Debug("Faro logs are nil, skipping")
+		}
+	} else {
+		r.settings.Logger.Debug("Logs consumer not registered, skipping")
+	}
+
+	if len(errors) > 0 {
+		r.settings.Logger.Error("Failed to process Faro payload", zap.Any("payload", payload), zap.Any("errors", errors))
```

Given that clients may re-send, should we fail fast? IE if traces fails, do we _want_ to try to send logs?

---

I think yes - I'm making this fail fast so not to waste time when things go wrong.